pkg/connector/kairosdb.go

```diff
@@ -0,0 +1,494 @@
+	}
+	defer response.Body.Close()
+
+	if err = kairosdbCheckBackendResponse(response); err != nil {
+		return nil, fmt.Errorf("kairosdb[%s]: invalid HTTP backend response: %s", connector.name, err)
+	}
+
+	data, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, fmt.Errorf("kairosdb[%s]: unable to read HTTP response body: %s", connector.name, err)
+	}
+
+	var JSONresponse map[string][]MetricsQueryResponse
+	if err = json.Unmarshal(data, &JSONresponse); err != nil {
+		return nil, fmt.Errorf("kairosdb[%s]: unable to unmarshal JSON data: %s", connector.name, err)
+	}
+
+	if resultSeries, err = kairosdbExtractPlots(query, connector.series, JSONresponse["queries"]); err != nil {
+		return nil, fmt.Errorf("kairosdb[%s]: unable to extract plot values from backend response: %s", connector.name, err)
+	}
+
+	// return nil, fmt.Errorf("kairosdb[%s]: not implemented, yet.", connector.name)
+
+	return resultSeries, nil
+}
+
+// Refresh triggers a full connector data update.
+func (connector *KairosdbConnector) Refresh(originName string, outputChan chan<- *catalog.Record) error {
+	type MetricsQuery struct {
+		Start_relative map[string]string   `json:"start_relative"`
+		Metrics        []map[string]string `json:"metrics"`
+	}
+
+	var JSONmetrics map[string][]string
+	var JSONquery map[string][]map[string][]struct {
+		Name string              `json:"name"`
+		Tags map[string][]string `json:"tags"`
+	}
+
+	httpTransport := &http.Transport{
+		Dial: (&net.Dialer{
+			// Enable dual IPv4/IPv6 stack connectivity:
+			DualStack: true,
+			// Enforce HTTP connection timeout:
+			Timeout: time.Duration(connector.timeout) * time.Second,
+		}).Dial,
+	}
+
+	if connector.insecureTLS {
+		httpTransport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+	}
+
+	httpClient := http.Client{Transport: httpTransport}
+
+	request, err := http.NewRequest("GET", strings.TrimSuffix(connector.URL, "/")+kairosdbURLMetricNames, nil)
+	if err != nil {
+		return fmt.Errorf("kairosdb[%s]: unable to set up HTTP request: %s", connector.name, err)
+	}
+
+	request.Header.Add("User-Agent", "Facette")
+	request.Header.Add("X-Requested-With", "KairosdbConnector")
+
+	response, err := httpClient.Do(request)
+	if err != nil {
+		return fmt.Errorf("kairosdb[%s]: unable to perform HTTP request: %s", connector.name, err)
+	}
+	defer response.Body.Close()
+
+	if err = kairosdbCheckBackendResponse(response); err != nil {
+		return fmt.Errorf("kairosdb[%s]: invalid HTTP backend response: %s", connector.name, err)
+	}
+
+	data, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return fmt.Errorf("kairosdb[%s]: unable to read HTTP response body: %s", connector.name, err)
+	}
+	if err = json.Unmarshal(data, &JSONmetrics); err != nil {
+		return fmt.Errorf("kairosdb[%s]: unable to unmarshal JSON data: %s", connector.name, err)
+	}
+
+	// REVIEW: whats a good interval to populate metrics from?
```

Comment:
Am 18.01.2015 um 13:41 schrieb Marc Falzon:
> In pkg/connector/kairosdb.go
> <https://github.com/facette/facette/pull/153#discussion-diff-23132988>:
> 
>> +	// REVIEW: whats a good interval to populate metrics from?
> 
> I guess we would like to retrieve all the metrics ever stored, so I
> suggest specifying StartAbsolute = 0. WDYT?
Hmm... I am still undecided.
It seems like a question of what the user expects.

Imagine, someone stored a metric and stopped feeding
2 years ago. So the user see the metric in the frontend,
but doesn't get any data unless he goes back in time
for a couple of years.

What do you think about make it configurable
as a reasonable compromise for now?